lib/Dialect/SV/Transforms/PrettifyVerilog.cpp

```diff
@@ -62,6 +71,214 @@ static bool isVerilogUnaryOperator(Operation *op) {
+};
+
+/// Split an assignment to a structure to writes of individual fields.
+bool PrettifyVerilogPass::splitStructAssignment(OpBuilder &builder,
+                                                hw::StructType ty, Value dst,
+                                                Value src) {
+  // Follow a chain of injects to find all fields overwritten and separate
+  // them into a series of field updates instead of a whole-structure write.
+  DenseMap<StringAttr, std::pair<Location, Value>> fields;
+  while (auto inj = dyn_cast_or_null<hw::StructInjectOp>(src.getDefiningOp())) {
+    // Inner injects are overwritten by outer injects.
+    // Insert does not overwrite the store to be lowered.
+    auto field = std::make_pair(inj.getLoc(), inj.newValue());
+    fields.try_emplace(inj.fieldAttr(), field);
+    src = inj.input();
+  }
+
+  // The origin must be either the object itself (partial update)
+  // or the whole object must be overwritten.
+  if (!isSelfWrite(dst, src) && ty.getElements().size() != fields.size())
+    return false;
+
+  // Emit the field assignments in the order of their definition.
+  for (auto &field : ty.getElements()) {
+    const auto &name = field.name;
+
+    auto it = fields.find(name);
+    if (it == fields.end())
+      continue;
+
+    auto &[loc, value] = it->second;
+    auto ref = builder.create<sv::StructFieldInOutOp>(loc, dst, name);
+    if (!splitAssignment(builder, ref, value))
+      builder.create<sv::PAssignOp>(loc, ref, value);
+  }
+  return true;
+}
+
+/// Split an assignment to an array element to writes of individual indices.
+bool PrettifyVerilogPass::splitArrayAssignment(OpBuilder &builder,
+                                               hw::ArrayType ty, Value dst,
+                                               Value src) {
+  // Follow a chain of concat + slice operations that alter a single element.
+  if (auto op = dyn_cast_or_null<hw::ArrayCreateOp>(src.getDefiningOp())) {
+    // TODO: consider breaking up array assignments into assignments
+    // to individual fields.
+    auto ty = hw::type_cast<hw::ArrayType>(op.getType());
+    if (ty.getSize() != 1)
+      return false;
+    APInt zero(std::max(1u, llvm::Log2_64_Ceil(ty.getSize())), 0);
+
+    Value value = op.inputs()[0];
+    auto loc = op.getLoc();
+    auto index = builder.create<hw::ConstantOp>(loc, zero);
+
+    auto field = builder.create<sv::ArrayIndexInOutOp>(loc, dst, index);
+    if (!splitAssignment(builder, field, value))
+      builder.create<sv::PAssignOp>(loc, field, value);
+    return true;
+  }
+
+  // TODO: generalise to ranges and arbitrary concatenations.
+  SmallVector<std::tuple<APInt, Location, Value>> fields;
+  while (auto concat =
+             dyn_cast_or_null<hw::ArrayConcatOp>(src.getDefiningOp())) {
+    auto loc = concat.getLoc();
+
+    // Look for a slice and an element:
+    // concat(slice(a, 0, size - 1), elem)
+    // concat(elem, slice(a, 1, size - 1))
+    if (concat.getNumOperands() == 2) {
```

Comment:
Pattern matching `concat(slice, elem)` , `concat(elem, slice)`, `concat(slice, elem, slice)` at PrettifyVerilog looks a bit dirty to me. Can we introduce `array_inject` in a similar way to `struct_inject` and move these pattern matching into canonilicazations? By doing this, I guess we can almost the same code as `spilitStructAssignmet`. WDYT?

Reply:
I've had the discussion with @darthscsi and it was decided that we should not have the op if we can represent it through concat/slice.

Reply:
Thanks, that's reasonable.